Adding 96 Nextbikes locations

gszathmari opened the pull request:
Adding support for 96 Nexbikes locations in the following cities:
- 10vorWien
- Al Sharjah
- Auckland
- Augsburg
- Bath
- Berlin
- Bielefeld
- Bietigheim-Bissingen
- Bochum
- Bottrop
- Cambridge
- Christchurch
- Dobrich
- Dortmund
- Dresden
- Dubai
- Duisburg
- Düsseldorf
- Essen
- Flensburg
- Frankfurt
- Friedrichshafen
- Gelsenkirchen
- Glasgow
- Gospić
- Grodzisk Mazowiecki
- Gütersloh
- Hamburg
- Hamm
- Hannover
- Heidelberg
- Herne
- Herrenberg
- Hoboken
- Hollabrunn
- Innsbruck
- Juchnowiec Kościelny
- Jūrmala
- Karlovac
- Karlsruhe
- Kent State University
- King Abdullah Economic City
- Konya
- Krems Ost
- Laa an der Thaya
- Leipzig
- Limassol
- Lublin
- Ludwigsburg
- Ludwigshafen
- Luzern
- Magdeburg
- Mannheim
- Marchfeld
- Mödling
- Mülheim a.d.R.
- München
- NeusiedlerSee
- Norderstedt
- Nürnberg
- ÖBB-Bahnhöfe
- Oberes Ybbstal
- Oberhausen
- Offenbach am Main
- Offenburg
- Piestingtal
- Pittsburgh
- Potsdam
- Quickborn
- Rīga
- Römerland
- Schwieberdingen
- Seferihisar
- Serfaus
- Šibenik
- Speyer
- Stirling
- St.Pölten
- Südheide
- Sursee
- Thermenregion
- Traisen-Gölsental
- Triestingtal
- Tübingen
- Tulln
- Tullnerfeld West
- Unteres Traisental
- Usedom
- Vaihingen an der Enz
- Wachau
- Waiblingen
- West Palm Beach Florida
- WienerWald
- Wr.Neustadt
- Würzburg
- Zagreb

eskerda commented:
PR under review

eskerda commented:
Where do names come from?

Usually we settle for just the name of the system (except on bicincitta, as far as I've checked) if none is known. Since city and country are supplied too the city name on the name of the network is redundant information (unless the system is known as such, like `Belfast Bikes`).

My recommendation will be to remove any country (and if possible city) from the network name:
* just `nextbike` is preferred over `nextbike Germany`
* `Leihradl` is good, unless it is known in Austria as `Leihradl nextbike`, then that is fine

About cities that have just one station, I am not sure if it makes sense to support them.

Make sure you run the tests as:
```
# Would run all tests (all networks), slow
sh runtests.sh
# Test data files for correctness
sh runtests.sh TestDataFiles
# Test nextbike
sh runtests.sh TestSystems.test_nextbike
```

Tests have detected that some instances in nextbike.json have lat / lng specified as strings instead of floats.

gszathmari commented:
The names are contructed as the following:

Sample response from https://nextbike.net/maps/nextbike-live.xml?domains=ch

```xml
<markers>
  <country lat="47.2569" lng="8.62976" zoom="9" name="nextbike Switzerland" hotline="+41415080800" domain="ch" country="CH" country_name="Switzerland" terms="http://www.nextbike.ch/media/AGB_CALU_2015.pdf" website="http://www.nextbike.ch/">
    <city uid="88" lat="47.1713" lng="8.10877" zoom="14" maps_icon="" alias="sursee" break="0" name="Sursee">
      <place uid="286233" lat="47.178162" lng="8.097114" name="Caritas Wohnen" spot="1" number="9001" bikes="3" bike_numbers="40181,40171,40190"/>
```

```python
        name = country['name'] + '-' + city['name']
        network['tag'] = slugify(name)
```

So `name` is `CH-Sursee`, which becomes `ch-sursee` after I slugify that

gszathmari commented:
Latitude and longitude also comes from https://nextbike.net/maps/nextbike-live.xml?domains=ch

```xml
<city uid="88" lat="47.1713" lng="8.10877" zoom="14" maps_icon="" alias="sursee" break="0" name="Sursee">
```

I took `lat` and `lng`

gszathmari commented:
I am going to update the pull request later with the `latitude` and `longitude` data in float format

gszathmari commented:
Actually I have built a quick 'n' dirty Nextbikes parser to convert the `nextbike-live.xml` into `nextbike.json`

Here is the script if you want to have a look on the logic
https://gist.github.com/gszathmari/bd7123b76374a0d74288

gszathmari commented:
I have fixed the lat/lon, could you please have a look again @eskerda ?

eskerda commented:
> Actually I have built a quick 'n' dirty Nextbikes parser to convert the nextbike-live.xml into nextbike.json
> Here is the script if you want to have a look on the logic https://gist.github.com/gszathmari/bd7123b76374a0d74288#file-nextbike-py

There's something similar in place at https://github.com/eskerda/pybikes/blob/master/extractors/nextbike.py . Not sure in which state I left it, though.

> I have fixed the lat/lon, could you please have a look again @eskerda ?

@gszathmari I'm afraid this PR will have to wait until 14th. I am going on vacation this Friday and there's much to work (meta wise) on this, as I do not like names coming directly from nextbike polluting the API. I am alright in merging by then and manually change the names

gszathmari commented:
Can I help you with the names? I am willing to amend my script to create the tags differently.

eskerda commented:
> Can I help you with the names? I am willing to amend my script to create the tags differently.

Usually we settle for just the name of the system (except on bicincitta, as far as I've checked) if none is known. Since city and country are supplied too the city name on the name of the network is redundant information (unless the system is known as such, like Belfast Bikes).

My recommendation will be to remove any country (and if possible city) from the network name:

* just nextbike is preferred over nextbike Germany
* Leihradl is good, unless it is known in Austria as Leihradl nextbike, then that is fine
* About cities that have just one station, I am not sure if it makes sense to support them.

gszathmari commented:
What would be a reasonable compromise here that can also be scripted? We have the following input from the xml feed to construct the JSON file:
```xml
  <country lat="50.7086" lng="10.6348" zoom="5" name="nextbike Germany" hotline="+493069205046" domain="de" country="DE" country_name="Germany" terms="http://www.nextbike.de/media/nextbike_AGB_de_en_03-2015.pdf" website="http://www.nextbike.de/">
    <city uid="1" lat="51.3415" lng="12.3625" zoom="14" maps_icon="" alias="leipzig" break="0" name="Leipzig">
```

eskerda commented:
Since http://www.nextbike.de/en/leipzig/ provides no further info about a name, just `nextbike` would be good.

Actually, I am thinking on allowing the definition of networks in pybikes without a name at all, in which case the name of the system will be taken

gszathmari commented:
So `name` should be fixed to `nextbike` in all cases as below?

```
        {
            "domain": "de",
            "tag": "nextbike-germany-leipzig",
            "meta": {
                "latitude": "51.3415",
                "city": "Leipzig",
                "name": "nextbike",
                "longitude": "12.3625",
                "country": "DE"
            },
            "city_uid": "1"
        },
        {
            "domain": "de",
            "tag": "nextbike-germany-tubingen",
            "meta": {
                "latitude": "48.5203",
                "city": "Tübingen",
                "name": "nextbike",
                "longitude": "9.05591",
                "country": "DE"
            },
            "city_uid": "101"
        }
```

eskerda commented:
Since this commit now https://github.com/eskerda/pybikes/commit/cb741b1dea6f1749926eaa51316cc2473cd54a4d there's no need to define a name when no name exists. It will get automatically set to whatever the provider has declared on its meta.

For tag generation, `{provider_name}-{city_name}` seems reasonable.

So your input would become:
```
        {
            "domain": "de",
            "tag": "nextbike-leipzig",
            "meta": {
                "latitude": "51.3415",
                "city": "Leipzig",
                "longitude": "12.3625",
                "country": "DE"
            },
            "city_uid": "1"
        },
        {
            "domain": "de",
            "tag": "nextbike-tubingen",
            "meta": {
                "latitude": "48.5203",
                "city": "Tübingen",
                "longitude": "9.05591",
                "country": "DE"
            },
            "city_uid": "101"
        }
```

gszathmari commented:
Okay I removed the `meta.name` property, and changed the `tag` as the following:
* in case of a Nextbike provider: `nextbike-` + `city.name`
* otherwise `country.name` + `city.name` from the XML feed

eskerda commented:
> Okay I removed the meta.name property

But for networks that have a know name we must have one!: `Belfast Bikes`, etc

gszathmari commented:
Okay I have added `meta.name` back

eskerda commented:
That's not what I meant.

There are systems that have name, others do not have one.

For instance, Belfast Bikes have a name. People know the belfast bikes system as "Belfast Bikes". However, Nextbike Leipzig does not have a name, people refer to it as just "nextbike" (or so I guess). For these, no name must be specified.

gszathmari commented:
I would keep "Nextbike Leipzig" because this is how the company refers to itself. If you visit their homepage on http://www.nextbike.de/en/leipzig/, the title also says its "Public bike rental in Leipzig - **nextbike Leipzig**"

Unfortunately I do not have any insight into the local nicknames of the individual Nextbike branches. However if you provide me a simple *if-then* logic I am happy to incorporate that into the script.

eskerda commented:
> I would keep nextbike Leipzig because this is how the company refers to itself. If you visit their homepage on http://www.nextbike.de/en/leipzig/, the title also says it is "Public bike rental in Leipzig - nextbike Leipzig"

To me the `Leipzig` in `nextbike Leipzig` is redundant, as the name of the city is already provided on the network. But many networks in citybikes are like this (call a bike, bicincitta, etc), so I guess both work.

The only insight I can provide is, given the names that come on the xml, guess which ones are proper names (Belfast Bikes) and which ones are not (nextbike Leipzig germany). As I said, this probably requires manual doing.

gszathmari commented:
Sounds good, let me know if I can help you with anything to merge the PR

On 3 September 2015 at 23:33, eskerda <notifications@github.com> wrote:

> I would keep nextbike Leipzig because this is how the company refers to
> itself. If you visit their homepage on http://www.nextbike.de/en/leipzig/,
> the title also says it is "Public bike rental in Leipzig - nextbike Leipzig"
>
> To me the Leipzig in nextbike Leipzig is redundant, as the name of the
> city is already provided on the network. But many networks in citybikes are
> like this (call a bike, bicincitta, etc), so I guess both work.
>
> The only insight I can provide is, given the names that come on the xml,
> guess which ones are proper names (Belfast Bikes) and which ones are not
> (nextbike Leipzig germany). As I said, this probably requires manual doing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eskerda/pybikes/pull/57#issuecomment-137591642>.
>



-- 
Gabor Szathmari
gszathmari@gmail.com

eskerda commented:
All done. Thanks @gszathmari !
